src/Draco.Compiler/Internal/Syntax/AbstractSyntaxTree.cs

```diff
@@ -0,0 +1,197 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Draco.Compiler.Internal.Syntax;
+
+internal abstract partial record class AbstractSyntaxTree
+{
+    public sealed record class Symbol() : AbstractSyntaxTree;
+
+    public abstract partial record class Decl : AbstractSyntaxTree
+    {
+        /// <summary>
+        /// A function declaration.
+        /// </summary>
+        public sealed partial record class Func(
+            Symbol Identifier,
+            ImmutableArray<Symbol> Params,
+            Symbol? ReturnType,
+            FuncBody Body) : Decl;
+
+        /// <summary>
+        /// A label declaration.
+        /// </summary>
+        public sealed partial record class Label(
+            Symbol Identifier) : Decl;
+
+        /// <summary>
+        /// A variable declaration.
+        /// </summary>
+        public sealed partial record class Variable(
+            Symbol Identifier,
+            Symbol? Type,
+            Expr Value) : Decl;
+    }
+
+    /// <summary>
+    /// A function body
+    /// </summary>
+    public partial record class FuncBody : AbstractSyntaxTree
+    {
+        /// <summary>
+        /// A block function body.
+        /// </summary>
+        public sealed partial record class BlockBody(
+            Expr.Block Block) : FuncBody;
+
+        /// <summary>
+        /// An in-line function body.
+        /// </summary>
+        public sealed partial record class InlineBody(
+            Expr Expression) : FuncBody;
+    }
+
+    public abstract partial record class Expr : AbstractSyntaxTree
+    {
+        /// <summary>
+        /// A block expression
+        /// </summary>
+        public partial record class Block(
+            ImmutableArray<Stmt> Statements,
+            Expr? Value) : Expr;
+
+        /// <summary>
+        /// A literal expression, i.e. a number, string, boolean value, etc.
+        /// </summary>
+        public sealed partial record class Literal(
+            Symbol Value) : Expr;
```

Comment:
Not sure if constants are symbols, I don't think they should be, but they should contain a symbol that represents their type